shell/platform/android/io/flutter/embedding/engine/systemchannels/TextInputChannel.java

```diff
@@ -0,0 +1,391 @@
+    channel.invokeMethod(
+        "TextInputClient.performAction",
+        Arrays.asList(inputClientId, "TextInputAction.search")
+    );
+  }
+
+  /**
+   * Instructs Flutter to execute a "send" action.
+   */
+  public void send(int inputClientId) {
+    channel.invokeMethod(
+        "TextInputClient.performAction",
+        Arrays.asList(inputClientId, "TextInputAction.send")
+    );
+  }
+
+  /**
+   * Instructs Flutter to execute a "done" action.
+   */
+  public void done(int inputClientId) {
+    channel.invokeMethod(
+        "TextInputClient.performAction",
+        Arrays.asList(inputClientId, "TextInputAction.done")
+    );
+  }
+
+  /**
+   * Instructs Flutter to execute a "next" action.
+   */
+  public void next(int inputClientId) {
+    channel.invokeMethod(
+        "TextInputClient.performAction",
+        Arrays.asList(inputClientId, "TextInputAction.next")
+    );
+  }
+
+  /**
+   * Instructs Flutter to execute a "previous" action.
+   */
+  public void previous(int inputClientId) {
+    channel.invokeMethod(
+        "TextInputClient.performAction",
+        Arrays.asList(inputClientId, "TextInputAction.previous")
+    );
+  }
+
+  /**
+   * Instructs Flutter to execute an "unspecified" action.
+   */
+  public void unspecifiedAction(int inputClientId) {
+    channel.invokeMethod(
+        "TextInputClient.performAction",
+        Arrays.asList(inputClientId, "TextInputAction.unspecified")
+    );
+  }
+
+  /**
+   * Sets the {@link TextInputMethodHandler} which receives all events and requests
+   * that are parsed from the underlying platform channel.
+   */
+  public void setTextInputMethodHandler(@Nullable TextInputMethodHandler textInputMethodHandler) {
+    this.textInputMethodHandler = textInputMethodHandler;
+  }
+
+  public interface TextInputMethodHandler {
+    // TODO(mattcarroll): javadoc
+    void show();
+
+    // TODO(mattcarroll): javadoc
+    void hide();
+
+    // TODO(mattcarroll): javadoc
+    void setClient(int textInputClientId, @NonNull Configuration configuration);
+
+    // TODO(mattcarroll): javadoc
+    void setEditingState(@NonNull TextEditState editingState);
+
+    // TODO(mattcarroll): javadoc
+    void clearClient();
+  }
+
+  /**
+   * A text editing configuration.
+   */
+  public static class Configuration {
+    public static Configuration fromJson(@NonNull JSONObject json) throws JSONException {
+      final Integer inputAction = inputActionFromTextInputAction(json.getString("inputAction"));
+      return new Configuration(
+          json.optBoolean("obscureText"),
+          json.optBoolean("autocorrect", true),
+          TextCapitalization.fromValue(json.getString("textCapitalization")),
+          InputType.fromJson(json.getJSONObject("inputType")),
+          inputAction,
+          json.optString("actionLabel")
+      );
+    }
+
+    private static Integer inputActionFromTextInputAction(@Nullable String inputAction) {
+      if (inputAction == null) {
```

Comment:
Should we also be returning null for empty strings?

Reply:
The existing logic appears to return EditorInfo.IME_ACTION_UNSPECIFIED for "". It's possible that that behavior is a mistake, but until we find out for sure I wouldn't want to alter the execution path.  That may be intentional.